File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque
def solution(begin, target, words):
    answer = 0
    q = deque()
    q.append([begin, 0])
    visited = [0] * len(words)
    
    while q:
        b, c = q.popleft()
        if b == target:
            answer = c
            break
        for i in range(len(words)):
            cnt = 0
            for j in range(len(words[0])):
                if not visited[i]:
                    if b[j] != words[i][j]:
                        cnt += 1
            if cnt == 1:
                visited[i] = 1
                q.append([words[i], c+1])       
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;
class Solution {
    public class Word{
        String word;
        int cnt;
        
        public Word(String word, int cnt){
            this.word = word;
            this.cnt = cnt;
        }
    }
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        Queue<Word> q = new LinkedList<>();
        q.add(new Word(begin, 0));
        boolean[] visited = new boolean[words.length];
        
        while(!q.isEmpty()){
            Word now = q.poll();
            if(now.word.equals(target)) {
                answer = now.cnt;
                break;
            }
            for(int i = 0; i < words.length; i++){
                if(visited[i]) continue;
                int cnt = 0;
                for(int j = 0; j < words[i].length(); j++){
                    if(words[i].charAt(j) == now.word.charAt(j)){
                        cnt++;
                    }
                }
                if(cnt == words[i].length() - 1){
                    visited[i] = true;
                    q.add(new Word(words[i], now.cnt + 1));
                }
            }
        }
        return answer;
    }
}